glib-macros/src/gclosure.rs

```diff
@@ -0,0 +1,220 @@
+            "`{}`: Field accesses are not allowed as is, you must rename it!",
+            name
+        );
+        Ok(Capture { name, alias, kind })
+    }
+}
+
+struct Closure {
+    captures: Vec<Capture>,
+    args: Vec<Ident>,
+    closure: syn::ExprClosure
+}
+
+impl syn::parse::Parse for Closure {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let mut captures = vec![];
+        if input.peek(Token![@]) {
+            loop {
+                captures.push(input.parse()?);
+                if input.peek(Token![,]) {
+                    input.parse::<Token![,]>()?;
+                    if !input.peek(Token![@]) {
+                        break;
+                    }
+                } else {
+                    break;
+                }
+            }
+        }
+        if !captures.is_empty() {
+            input.parse::<Token![=>]>()?;
+        }
+        let mut closure = input.parse::<syn::ExprClosure>()?;
+        if closure.asyncness.is_some() {
+            panic!("Async closure not allowed");
+        }
+        if !captures.is_empty() && closure.capture.is_none() {
+            panic!("Closure with captures needs to be \"moved\" so please add `move` before closure")
+        }
+        let args = closure.inputs.iter().enumerate().map(|(i, _)| {
+            Ident::new(&format!("____value{}", i), Span::call_site())
+        }).collect();
+        closure.capture = None;
+        Ok(Closure {
+            captures,
+            args,
+            closure
+        })
+    }
+}
+
+impl ToTokens for Closure {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let closure_ident = Ident::new("____closure", Span::call_site());
+        let values_ident = Ident::new("____values", Span::call_site());
+        let crate_ident = crate_ident_new();
+
+        let outer_before = self.captures.iter().map(|c| c.outer_before_tokens(&crate_ident));
+        let inner_before = self.captures.iter().map(|c| c.inner_before_tokens(&crate_ident));
+        let outer_after = self.captures.iter().map(|c| c.outer_after_tokens(&crate_ident, &closure_ident));
+
+        let arg_values = self.args
+            .iter()
+            .enumerate()
+            .map(|(index, arg)| {
+                let err_msg = format!(
+                    "Wrong type for argument {} in gclosure: {{:?}}",
+                    index
+                );
+                quote! {
+                    let #arg = ::core::result::Result::unwrap_or_else(
+                        #crate_ident::Value::get(&#values_ident[#index]),
+                        |e| panic!(#err_msg, e),
+                    );
+                }
+            });
+        let arg_names = &self.args;
+        let closure = &self.closure;
+
+        tokens.extend(quote! {
+            {
+                let #closure_ident = {
+                    #(#outer_before)*
+                    #crate_ident::closure::Closure::new_local(move |#values_ident| {
```

Comment:
This should probably get a `_local` and "normal" variant for the macro so both can be created

Reply:
Added a second macro for this.